keps/sig-network/3458-remove-transient-node-predicates-from-service-controller/README.md

```diff
@@ -95,6 +95,24 @@ concerns cluster ingress connectivity. Users of Kubernetes will also see a
 reduction in the amount of cloud API calls, for what concerns calls stemming
 from syncing load balancers with the Kubernetes cluster state.
 
+Addressing b) is not useful for ingress load balancing. A load balancer needs to
+know if the networking data plane is running fine and this is determined by the
+configured health check. Cloud providers define their own health check, and no
+one does the same. The following describes what the health check looks like on
+the three major public cloud providers:
+
+- GCP: probes port 10256 (Kube-proxy's healthz port)
+- AWS: if ELB; probes the first `NodePort` defined on the service spec
+- Azure: probes all `NodePort` defined on the service spec.
+  
+All clouds take an approach of trying to ascertain if traffic can be forwarded
+to the endpoint, which is a completely valid health check for load balancer
+services. There are drawbacks to all of these ways of doing - but cloud
+providers themselves are deemed best suited for what concerns: determining what
+is the best mechanism to use for their load balancers / cloud's mode of
+operation. Their mechanism is beyond the scope of this KEP, i.e: this KEP does
+not attempt to "align them".
+
```

Comment:
L269: I wouldn't say it's N/A - the test is not needed because it's in-memory feature, but it requires some comment.

L275: That is universally true :) The question is what can possibly go wrong.
My examples:
- cluster has a gazillion of not-ready nodes we try to add them on upgrade and that fails for whatever reason (cloud-provider rejects the request because there are too many of those in in a single request)
- there is a bug in implementation of some case and ccm crashes
etc.

L279: Do we have some metrics for cloud-provider calls?
 Having a metrics reporting success/error for calls to cloud-providers would be super useful here.
 If we don't have them, can we add an AI to this KEP to add them?
[OK - I see new metrics being mentioned below - let' mention the second one here too - it's pretty useful.]

L283: I still  would like to see some manual test to be run before promoting to beta for sanity check (a cluster with both ready and non-ready nodes is being upgraded). 


L315: I would rather focus on the fact of not broken connections when nodes are becoming not ready.

L323: that's not really an SLO - I'm fine with saying no SLO, because we can have a reasonable SLI

L335: let's put a metric name there (you proposed adding a metric for that above)

L347: dependencies are about "service dependencies" (things that needs to work to make the feature work, like some other service, cloud-provider, external DNS, etc.)
This change doesn't have any dependency.

Reply:
> L275: That is universally true :) The question is what can possibly go wrong.
> My examples:

I can add that, but it's just as generic. 

> L283: I still would like to see some manual test to be run before promoting to beta for sanity check (a cluster with both ready and non-ready nodes is being upgraded).

I can't manually test this. I don't have access to development environments where I can benefit from testing Kubernetes HEAD. And I need a Kubernetes environment running on a cloud, so Kind doesn't apply to this. If someone else working on a Kubernetes distribution could help, I'd be happy to advise. 

> L315: I would rather focus on the fact of not broken connections when nodes are becoming not ready.

That's much more difficult for someone to do, I think. The question asks "for someone using the feature", so I assume a user. We can try adding an E2E disruption test that validates something like this?  

> L323: that's not really an SLO - I'm fine with saying no SLO, because we can have a reasonable SLI

OK, I will mention it as an SLI then and put N/A on SLO 

> L335: let's put a metric name there (you proposed adding a metric for that above)

It might be difficult to compare though given that the metric doesn't exist on existing versions. The event will exist before and after. I will add the metric name though

Reply:
> If someone else working on a Kubernetes distribution could help, I'd be happy to advise.

I can help - kube-up still works on GCP :)

Reply:
> I can add that, but it's just as generic.

Still - I remember such failure mode in the past. This change can in theory re-expose it if we didn't fix it correctly.
Most of the failure modes we believe won't materialize, but enumerating them helps for making them even less probable, because we intentially start thinking about them.
Please add and think about others.

> I can't manually test this. I don't have access to development environments where I can benefit from testing Kubernetes HEAD. And I need a Kubernetes environment running on a cloud, so Kind doesn't apply to this. If someone else working on a Kubernetes distribution could help, I'd be happy to advise.

From my perspective it doesn't really matter who will do the testing.
But the testing *HAVE TO HAPPEN* - we can't be releasing things without even checking if these work.

@thockin - may I ask you to ensure that such tests happened (and the KEP will be updated decrining the test done) before the feature-gate will be switched to Beta in k/k?


> That's much more difficult for someone to do, I think. The question asks "for someone using the feature", so I assume a user. We can try adding an E2E disruption test that validates something like this?

Yes - it's about a user. But as a user, I don't really care (or may not even be aware) what is exposed from LB (as this is something I don't own - it's an underlying infrastructure). What I see though is whether connections to my apps are non-gracefully breaking.

Reply:
Yes, we can pair up and design a test for this